### PR TITLE
Fix DeviceContext bug

### PR DESCRIPTION
**Pull request by mlazos**
Fixes https://github.com/pytorch/pytorch/issues/133666

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133137
* #133135
* #133136
* #133134
* #133133
* #133132
* #133131
* __->__ #133729
* #133130

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @xmfan

### CODE REVIEW

File: torch/utils/_device.py [hunk 1]
```diff
@@ -65,12 +66,36 @@ def __enter__(self):
         global CURRENT_DEVICE
         self.old_device = CURRENT_DEVICE
         CURRENT_DEVICE = self.device
-        return super().__enter__()
+        # We need to put the device at the bottom of the stack
```
Comment (bdhirsh) [inline torch/utils/_device.py:69]: hmm cc @albanD do you mind reviewing this too?

In particular - I'm not sure which implementation we want, between:

(1) this one (invariant that the `DeviceContext` mode only ever appears once on the stack)

(2) treat the mode more like an "ordinary" mode: you can push more than one of them on the stack, and the top-most mode will always take precedence. E.g. if you push `context1` onto the stack, then `context2`, `context2` will shadow the `context1`. when `context2` is popped later, `context1` applies.

Maybe related question for @mlazos - do you think (2) would make it easier to support this mode in dynamo without extra special handling?

Comment (bdhirsh) [reply torch/utils/_device.py:69]: Talked offline - it seems like given the fact that `set_default_device(None)` is supposed to pop the stack, we fundamentally cannot treat this mode as an "ordinary" mode (it's push/pop are not guaranteed to happen in the same order compared to other modes)

File: torch/utils/_device.py [hunk 2]
```diff
@@ -65,12 +66,36 @@ def __enter__(self):
...
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         global CURRENT_DEVICE
         CURRENT_DEVICE = self.old_device
-        return super().__exit__(exc_type, exc_val, exc_tb)
+        cur_stack = []
+        # Invariant: there should only be one DeviceContext on the stack at any time
+        # search the stack for the device context and pop it, restore previous stack
+        # entries popped while searching
+        for _ in range(_len_torch_function_stack()):
```
Comment (bdhirsh) [inline torch/utils/_device.py:90]: shouldn't we have an invariant that the bottom-most mode **must** be our `DeviceContext`? (So in particular, we can also assert only the last mode that we pop is of that type?)

**Review by bdhirsh — APPROVED** [review]

File: torch/utils/_device.py (hunk 2 above)
Comment (mlazos) [reply torch/utils/_device.py:90]: Yeah I'll add this assertion, it was missing in the previous impl!

**Pull request closed without merging**